### DISCUSSION TITLE
Extension VSIX to extend another VSIX

### DISCUSSION THREAD

**Discussion by cjdutoit** [VsixCommunity/Community.VisualStudio.Toolkit#358] [Q&A]
I need some help / direction.  Just started playing with the VS Extensions and I would like to create my own project that will extend functionality of another VSIX that is installed.

I started looking at sample extensions and landed on [Add New File](https://marketplace.visualstudio.com/items?itemName=MadsKristensen.AddNewFile)

Looking at the install location for this extension, I can see a **Templates** folder:
%LocalAppData%\Microsoft\VisualStudio\17.0_dae60504Exp\Extensions\Extensions-17.0_dae60504\qhs1xiqh.qqq\Templates

![image](https://user-images.githubusercontent.com/797750/176037920-53c2efde-4220-492d-b28d-c3c59a6cbaed.png)

What I would like to do is create my own extension, that could add my own custom templates inside the template folder above.  

So my questions:
1) Is there a way that I can check if the above extension is installed and if not prevent my extension from installing
2) If the extension is installed, is there a nice way to get the location so I can target the template folder for my custom extensions.

Any help / ideas would be greatly appreciated.

**Comment by reduckted** [comment]
> 1. Is there a way that I can check if the above extension is installed and if not prevent my extension from installing

Take a look at the `IVsExtensionManager` interface. It's not part of the SDK, but you can get an instance of the service and then call its methods using reflection.
https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2013/dd884736(v%3Dvs.120)

You can get an instance of it by defining this interface:
```cs
[Guid("316F4DE6-3CA4-4f0d-B003-962D28F65238")]
public interface SVsExtensionManager {}
```
and then use that interface to resolve the service:
```cs
object extensionManager = await VS.GetServiceAsync<SVsExtensionManager, object>();
```

> 2. If the extension is installed, is there a nice way to get the location so I can target the template folder for my custom extensions.

The [`IVsExtensionManager.GetInstalledExtension`](https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2013/dd884824(v=vs.120)) method returns an `IInstalledExtension`. It has an [`InstallPath`](https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2013/dd884737(v=vs.120)) property. Again, it's not part of the SDK, so you'll need to use reflection to access it. I can provide an example of how to access it using reflection if you like.

**Reply by cjdutoit** [reply]
Thanks @reduckted , I will have look and get back to you if I have further questions.
